okhttp/src/main/java/okhttp3/EventListener.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (C) 2017 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3;
+
+import java.net.InetAddress;
+import java.util.List;
+
+/**
+ * Observes, modifies, and potentially short-circuits requests going out and the corresponding
+ * responses coming back in. Typically interceptors add, remove, or transform headers on the request
+ * or response.
+ */
+public class EventListener {
```

Comment:
I think with the factory, the request is simpler. Without a factory, the question about reentrance is scoped to a call parameter, which makes my head hurt to explain :P On that note, I think since there's a factory now, the call parameters probably need to be squashed (unless the call parameters refer to a "network request")

Back to reentrance.. for the same call, would we ever see one of these callbacks twice? Which callbacks are optional? Which callbacks obviate others?

For example, connection callbacks are asymmetric wrt a call. We'd expect a start, end and abend to be on different calls (due to keepalives etc). However, a fetch start and end seems something that we can guarantee is one-one with a call. At least factory.create, end would be something we could guarantee.

The other thing is which callback would the exception raised from that connection break go to? Would a wrapped variant of that exception go somewhere else (with a root cause of the connection exception)? Mainly interested in the state diagram, knowing the whole design is a work in progress.